File: src/main/scala/codechicken/multipart/api/part/TNormalOcclusionPart.scala
package codechicken.multipart.api.part

import codechicken.multipart.api.NormalOcclusionTest
import net.minecraft.util.math.shapes.{IBooleanFunction, VoxelShape, VoxelShapes}

/**
 * Provides standard bounding box based occlusion testing.
 * If any two parts have overlapping bounding boxes, the test fails
 */
trait TNormalOcclusionPart extends TMultiPart {
    /**
     * Return a list of normal occlusion boxes
     */
    def getOcclusionShape: VoxelShape

    override def occlusionTest(npart: TMultiPart): Boolean =
        NormalOcclusionTest.test(this, npart) && super.occlusionTest(npart)
}

File: src/main/java/codechicken/multipart/api/part/TNormalOcclusionPart.java
package codechicken.multipart.api.part;

import codechicken.multipart.api.NormalOcclusionTest;
import net.minecraft.world.phys.shapes.VoxelShape;

/**
 * Provides standard bounding box based occlusion testing.
 * If any two parts have overlapping bounding boxes, the test fails
 */
public interface TNormalOcclusionPart extends TMultiPart {

    VoxelShape getOcclusionShape();

    @Override
    default boolean occlusionTest(TMultiPart npart) {
        return NormalOcclusionTest.test(this, npart);
    }
}
